Separate command and options

shadowhand opened the pull request:
Having separate value objects for options makes it easier to understand
what options are available to a command and simplifies the validation
process.

This change was implemented as separate classes so that usage can be
upgraded incrementally and version 2.0 can remove the deprecated things.

shadowhand committed:
Separate command and options

Having separate value objects for options makes it easier to understand
what options are available to a command and simplifies the validation
process.

This change was implemented as separate classes so that usage can be
upgraded incrementally and version 2.0 can remove the deprecated things.

shadowhand commented on composer.json:
```diff
@@ -13,7 +13,9 @@
-        "php": ">=5.5"
+        "php": ">=5.5",
+        "equip/assist": "^0.1.2",
+        "equip/data": "^2.3"
```
Should we separate the the immutable traits from this package to avoid version conflicts?

elazar replied:
Not sure I follow. Why would the immutable traits not being separate cause version conflicts?

shadowhand replied:
`equip/data` would depend on `equip/immutable` and this version would depend on `equip/immutable` only.

shadowhand replied:
The goal is to prevent packages that depend on `equip/data` from having conflicts with this package.

elazar replied:
Are we anticipating those sorts of conflicts would happen often enough to warrant the overhead that comes from maintaining an entirely separate repository?

shadowhand replied:
I'm not a fan having repository/entity stuff in the same package as immutable traits.

elazar replied:
That doesn't answer my question. Do you have specific use cases in mind wherein we would use immutable traits outside of repositories and entities? If so, do you anticipate conflicts happening with sufficient frequency to justify maintaining yet another repository?

shadowhand replied:
This package is an example... it uses the `ProtectedValueObjectTrait` but no repository interfaces.

elazar replied:
With the changes in this PR it does, yes. I [noted separately](https://github.com/equip/command/pull/6#issuecomment-205863177) that I'm not convinced of the value of this PR over the functionality it would deprecate.

shadowhand replied:
https://github.com/equip/immutable/pull/1 here's what it would look like.

elazar replied:
What it could look like isn't difficult to visualize, but that doesn't address the points I expressed above. How many use cases are we likely to encounter where we'd want to use the immutable traits apart from the repository and entity implementations, even assuming this repository did serve as a solitary example? I'm not convinced it's worth the trouble of separation and long-term maintenance versus pulling in a bit more code from api-commands than might be needed for this particular use case.

shadowhand replied:
Anywhere you want an immutable value object without entity/repo stuff.

shadowhand replied:
This could be really useful for notifications too.

elazar replied:
> Anywhere you want an immutable value object without entity/repo stuff.

Well, yeah. I included that in the question when I asked it. This doesn't answer it. I'm looking for specifics.

> This could be really useful for notifications too.

How so? Would we not have entities or repositories for notifications?

shadowhand commented on tests/AbstractCommandTest.php:
```diff
@@ -0,0 +1,140 @@
+
+use PHPUnit_Framework_TestCase as TestCase;
+
+class AbstractCommandTest extends TestCase
```
This is not modified, just moved.

shadowhand commented:
@equip/contributors CR please!

elazar commented:
> Having separate value objects for options makes it easier to understand what options are available to a command and simplifies the validation process.

Not sure I see how these changes accomplish what's described above. Can you give specific use cases that these changes would improve over the existing code?

shadowhand commented:
@elazar for example if we have a command that allows for some options that are not required and do not have defaults (say, "employee code") then having a value object makes it much more clear that this option exists.

Currently in many of our commands we are not sure of what all the possible options are, so having a value object makes a source of truth for these scenarios. It also makes the command implementation easier and separates concerns better, so that the command objects can contain less option checking.

shadowhand commented:
The code in this scenario would look something like:

``` php
final class CreateUserOptions extends Options
{
    protected $first_name;
    protected $last_name;
    protected $email;
    protected $password;
    protected $phone_number;
    protected $employee_code;
    protected $role = UsersRepository::ROLE_EMPLOYEE;
}
```

Edit: updated to match PR changes.

shadowhand commented:
@equip/contributors anyone else have thoughts about this?

elazar commented:
I think representing required options within `OptionsTrait` may be a flawed approach. It's possible two commands could take the same set of options, but require different options in that set (e.g. creating a user, where we'd want to require an e-mail address, versus updating a user, where we don't necessarily need an e-mail address unless it's being changed). By coupling requirements to the options themselves (versus the commands using them), we lose the ability to support this use case without adding further complexity.

I think having value objects for options is a good idea, but I also think it may be possible to implement it satisfactorily without deprecating existing code or breaking backward compatibility. For example, we could provide a generic implementation of `OptionsInterface` and, if an array is passed in for options, simply wrap it in an instance of that generic class within `AbstractCommand`. This leaves the possibility open for us to change the signature of `withOptions()` in the future to type hint against `OptionsInterface` in a future release and to simply recommend switching to use of `OptionsInterface` implementations in the interim.

shadowhand commented:
@elazar any further thoughts on this? I've updated my example code and switched to using an abstract class for Options (and no interface) because it's a lot more convenient.

elazar commented:
@shadowhand I don't follow: how is it more convenient to have no interface? This PR seems to be moving further away from design by contract and favoring composition over inheritance. I can see having a stock implementation of said interface, but I don't see why the command and options classes can't be implementations of an interface to allow for alternative implementations.

shadowhand commented:
@elazar YAGNI. The classes are so simple and high level that I don't see why anyone would need the interfaces. Type hinting against a `CommandInterface` gets you nothing, since every command implementation is totally unique and not swappable. There _might_ be some value in an `OptionsInterface` but I don't want to add it unnecessarily. It is trivial to add if someone actually requests it.

ameech commented:
This looks reasonable @shadowhand. Is this something we still want to move forward with?

shadowhand commented on src/Command.php:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Get a copy with new options.
+     *
+     * @param Options $options
```
Corrected this type hint, as per [this Scrutinizer issue](https://scrutinizer-ci.com/g/equip/command/inspections/fd08a0c3-1f33-43ae-afdc-a229929068da/issues/files/src/Command.php?status=new&orderField=path&order=asc&honorSelectedPaths=0).

shadowhand commented:
@ameech @elazar check this again please? I'd like to move forward with this.

elazar commented on src/Options.php:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @since 1.3.0
+ */
+abstract class Options
```
No immutable mutators (i.e. `with*()` methods)? If not, that's fine, just curious.

shadowhand replied:
No, because options cannot be modified after construction.

elazar commented on CHANGELOG.md:
```diff
@@ -7,6 +7,11 @@ This project adheres to [Semantic Versioning](http://semver.org/).
+## 1.3.0 - 2016-04-05
+
+- Added `Command` and `Options` abstract classes
+- Deprecated `AbstractCommand` and `CommandInterface`
```
We may want to have a separate document noting intended migration paths for deprecations like these?

shadowhand replied:
What would you suggest?

elazar replied:
MIGRATING.md, maybe? In the form, "if you're upgrading to this version, you'll need to modify any existing code in this way."

shadowhand committed:
Add MIGRATING document

shadowhand commented on MIGRATING.md:
```diff
@@ -0,0 +1,113 @@
+Migrating
```
@elazar document added.

elazar replied:
Just the one question on phrasing. Otherwise, looks good.

shadowhand replied:
Fixed.

elazar commented on MIGRATING.md:
````diff
@@ -0,0 +1,113 @@
+```
+
+All valid options are defined as `protected` class properties. Optional values
+can have default properties.
````
Should this not read "Optional properties can have default values?"

shadowhand replied:
Probably. 🙈

shadowhand replied:
Fixed.

elazar commented:
👍 

[![Approved with PullApprove](https://img.shields.io/badge/pullapprove-approved-brightgreen.svg)](https://pullapprove.com/equip/command/pull-request/6/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=equip/command)

shadowhand commented:
Released in 1.3.0.
